### PR TITLE
stop calling ListSteps when using pooled clusters

### PR DESCRIPTION
**Pull request by davidmarin**
Currently, our pooling code makes the EMR API call `ListSteps` on every eligible cluster, which can be expensive in terms of API calls (see #2159, which this fixes, for details).

So that two runners don't both submit a job to a cluster at the same time, there is a "locking" mechanism which uses S3. The current code includes both the cluster ID and the number of steps in the cluster in the URI of the lock file, so that a lock will be implicitly released once steps are added to the cluster. Instead, runners now explicitly release locks once their step(s) start running (fixing #2162), so we can just use the cluster ID.

We also ignored the limit on total number of steps in very old AMIs, in favor of telling the user pooling might not work properly on them, and trusting that the race case where a cluster has pending steps but is still in the `WAITING` stage (which seems to last about 10 seconds) will be prevented by cluster locking.

We also now hard-coded cluster lock expiry to last one minute, since we're just trying to cover the time between step submission and when the cluster starts running the steps.

Before v0.7.3 is released, I'm planning to replace the S3 "locking" mechanism with EMR tags, so I put a little less effort into unit-testing the locking mechanism than I otherwise would. I did hand-test pooling, and it worked as expected.

### CODE REVIEW

File: mrjob/emr.py [hunk 1]
```diff
@@ -920,6 +940,8 @@ def cleanup(self, mode=None):
             except Exception as e:
                 log.exception(e)
 
+        # TODO: otherwise, cancel any steps we submitted
```
Comment (88manpreet) [inline mrjob/emr.py:943]: Would it makes sense to add issue-id where it is supposed to be fixed?

File: mrjob/emr.py [hunk 2]
```diff
@@ -2590,26 +2557,24 @@ def _find_cluster(self, num_steps=1):
...
-                    self.fs.s3, self._lock_uri(cluster_id, cluster_num_steps),
-                    self._opts['cloud_fs_sync_secs'], self._job_key)
-                if status:
-                    log.debug('Acquired lock on cluster %s', cluster_id)
+                    len(cluster_id_list),
+                    '' if len(cluster_id_list) == 1 else 's',
+                    ': ' if cluster_id_list else '',
+                    ', '.join(reversed(cluster_id_list))))
+
+            # TODO: why not just try locking every cluster in the list?
```
Comment (88manpreet) [inline mrjob/emr.py:2570]: Same here?

**Comment by 88manpreet** [comment]
+1

File: mrjob/emr.py (hunk 1 above)
Comment (davidmarin) [reply mrjob/emr.py]: good point, done

File: mrjob/emr.py (hunk 2 above)
Comment (davidmarin) [reply mrjob/emr.py]: good point, done

**Comment by 88manpreet** [comment]
+1

**Pull request merged**